PWA first tutorial: first 3 parts

estelle opened the pull request:
Set up for intro level PWA tutorial including:

* static HTML and CSS
* JavaScript functionality - fully functional app, but not a PWA
* Manifest file: including appearance, definition, and iconography

Addition sections are currently being worked on.

estelle committed:
PWA: intro to basic tutorial

estelle committed:
PWA: intro to basic tutorial

estelle committed:
typo

estelle committed:
typo

estelle committed:
enhance headings

estelle committed:
enhance headings

estelle committed:
tweaks

estelle committed:
tweaks

estelle committed:
PWA: start of basic HTML page

estelle committed:
PWA: start of basic HTML page

estelle committed:
edits

estelle committed:
edits

estelle committed:
temporary edits

estelle committed:
temporary edits

estelle committed:
change from cat to cycle

estelle committed:
change from cat to cycle

estelle committed:
addition of cycle tracker app

estelle committed:
addition of cycle tracker app

estelle committed:
write explanation of HTML for PWA

estelle committed:
write explanation of HTML for PWA

estelle committed:
no clue

estelle committed:
no clue

estelle committed:
typos

estelle committed:
typos

estelle committed:
reorder content

estelle committed:
reorder content

estelle committed:
add js tut

estelle committed:
add js tut

estelle committed:
complete step1: html file

estelle committed:
complete step1: html file

estelle committed:
assets for step1

estelle committed:
assets for step1

estelle committed:
tweaks

estelle committed:
tweaks

estelle committed:
DL layout

estelle committed:
DL layout

estelle committed:
require dates, add past periods

estelle committed:
require dates, add past periods

estelle committed:
Step 2: the app.js file

estelle committed:
Step 2: the app.js file

estelle committed:
QC: html

estelle committed:
QC: html

estelle committed:
change div to section

estelle committed:
change div to section

estelle committed:
finish html section, hopefully

estelle committed:
finish html section, hopefully

estelle committed:
link text case

estelle committed:
link text case

estelle committed:
link text case

estelle committed:
link text case

estelle committed:
minor

estelle committed:
minor

estelle committed:
javascript lesson

estelle committed:
refinement of html

estelle committed:
refinement of javascript

estelle committed:
landing page

estelle committed:
manifest

estelle committed:
manifest

estelle committed:
intro to steps paragraph

estelle committed:
identity

estelle committed:
presentation and identity for the manifest

estelle committed:
Add intro to list of tutorials

estelle committed:
Add repo & series path. list of tut components

estelle committed:
repo and nav

estelle committed:
JS page tweaks

estelle committed:
Manifest page tasks, repo, nav, and presentation section

estelle committed:
appearance section

estelle committed:
rewrite appearance

estelle committed:
finished manifest draft. yes, it will change.

estelle committed:
links

estelle committed:
change file names and directory structure

estelle committed:
remove old files

estelle committed:
icongraphy and file names

estelle committed:
remove index so we can PR

estelle committed:
Merge branch 'main' into pwa1

estelle committed:
delete files not in this PR

estelle committed:
delete files not in this PR

estelle committed:
fix github issues

estelle committed:
@#$%

estelle committed:
manual prettier

estelle committed:
remove file from this PR

estelle committed:
URL: docs instead of Docs?

estelle committed:
Docs to docs

estelle committed:
flaws

estelle committed:
links to github instead of local

estelle committed:
tweaks

estelle committed:
Merge branch 'main' into pwa1

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md:
```diff
@@ -0,0 +1,351 @@
+
+## Up next
+
+At it's core, a PWA is a web application that is progressively enhanced to work offline. Now that we have a fully functional web application, we add the features required to convert into a PWA, including the [manifest file](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Manifest_file), [secure connection](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Secure), and [service worker](/en ...
```
```suggestion
At its core, a PWA is a web application that can be installed is progressively enhanced to work offline. Now that we have a fully functional web application, we add the features required to convert it to a PWA, including the [manifest file](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Manifest_file), [secure connection](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Secure), and [service worker](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Service_worker).
```

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,227 @@
+
+A PWA manifest file is a JSON file that provides information about the features of that app to make it look and behave like a native app when installed on the user's device. The manifest contains metadata for your app, including its name, icons, and presentational directives.
+
+The manifest file contains information about the PWA's identity, presentation, and iconography. While the only requirement is to include a name using the `name` or `short_name` key, you should also include icon information, the URL to be opened when the PWA is launched, and type of application viewport in which the PWA should be viewed.
```
```suggestion
While the only requirement is to include a name using the `name` or `short_name` key, you should also include icon information, the URL to be opened when the PWA is launched, and type of application viewport in which the PWA should be viewed.
```

This first sentence seems like a repeat of the previous one.

captainbrosset replied:
It depends what you mean by requirement. Will wrote a guide about PWA installation and it lists the manifest members that are required to install a PWA on a device: https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Guides/Making_PWAs_installable#required_manifest_members

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
````diff
@@ -0,0 +1,227 @@
+}
+```
+
+Before saving the manifest file and linking to it from our HTML file, let's discuss we can develop a still brief but more informative JSON object to define the identity, presentation, and iconography of the PWA. Yes, the above would work, but let's discuss a few more keys that enable manifest files to better define the appearance of our PWA.
````
```suggestion
Before saving the manifest file and linking to it from our HTML file, we can develop a still brief but more informative JSON object to define the identity, presentation, and iconography of the PWA. Yes, the above would work, but let's discuss a few more members that enable manifest files to better define the appearance of our PWA.
```

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,227 @@
+
+## App identity
+
+To identity your PWA the JSON must include a `name` or `short_name` key, or both, to define the PWA name. It can also include a `description`.
```
```suggestion
To identity your PWA the JSON must include a `name` or `short_name` member, or both, to define the PWA name. It can also include a `description`.
```

Top-level properties of a web app manifest are called members. We've been using that name in other PWA docs on MDN. It'd be good to stick to it consistently if we can.

captainbrosset reviewed (commented):
I'm not fully done with the review, but since I'm done with my day, I thought I would post my comments now, giving you some time to take a look at them, and I'll then go back to the review tomorrow, or a bit later this week.
Thanks for the work on this so far, this looks really good.

estelle committed:
Apply suggestions from code review

Co-authored-by: Patrick Brosset <patrickbrosset@gmail.com>

estelle commented on files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/index.md:
````diff
@@ -0,0 +1,240 @@
+</section>
+```
+
+You can try the [static period tracker shell](https://mdn.github.io/pwa-examples/intro/html_and_css) and view the [period tracker HTML and CSS source code](https://github.com/mdn/pwa-examples/tree/master/intro/html_and_css) on GitHub.
````
good catch!

estelle committed:
edits from review (start)

estelle committed:
explanation of the css

estelle committed:
updates per PR review

estelle committed:
manifest, and unifying the app name

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+
+A PWA manifest file is a JSON file that provides information about the features of that app to make it look and behave like a native app when installed on the user's device. The manifest contains metadata for your app, including its name, icons, and presentational directives.
+
+While according to the spec, all of the manifest members, or keys, are optional, some are browsers, operating systems, and app distributors have requirements, with a manifest needing to include specific keys for a web app to be a PWA. By including a name using the `name` or `short_name` key, icons that meet the minimum requirements, the URL to be opened when the PWA is launched, and type of appli ...
```
```suggestion
While according to the spec, all of the manifest members, or keys, are optional, some browsers, operating systems, and app distributors have requirements to include specific keys for a web app to be a PWA. By including a name using the `name` or `short_name` key, icons that meet the minimum requirements, the URL to be opened when the PWA is launched, and type of application viewport in which the PWA should be viewed, your app will meet the manifest requirements of a PWA.
```

Feel free to change this suggestion, but the original sentence sounded a bit odd to me.

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+
+While according to the spec, all of the manifest members, or keys, are optional, some are browsers, operating systems, and app distributors have requirements, with a manifest needing to include specific keys for a web app to be a PWA. By including a name using the `name` or `short_name` key, icons that meet the minimum requirements, the URL to be opened when the PWA is launched, and type of appli ...
+
+A minimalist manifest file for our menstrual cycle tracking app could look like this:
```
I would be good to link to Will's article about the required members, since that's the place that's most likely to be up to date.

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+      "sizes": "512x512"
+    }
+  ],
+  "display": "browser"
```
I think that, today, on all OS and devices, the `display:browser` does **not** make a web app a PWA. To be tested, but I think browsers won't offer the button to install the app if this display mode is used. So, even if you have a section about this later, it might make sense to use the `standalone` display mode here instead. Just for illustration purposes.

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+- [`display`](/en-US/docs/Web/Manifest/display)
+  - : Controls the app's display mode including `fullscreen`, `standalone`, which displays the [PWA as a standalone application](/en-US/docs/Web/Progressive_web_apps/How_to/Create_a_standalone_app), `minimal-ui`, which is similar to a standalone view but with UI elements for controlling navigation, and `browser`, which opens the app in a regular browser view.
+
+There is also an `orientation` key to define the PWA's default orientation as `portrait` or `landscape`. As our app works either way, we'll omit this key.
```
I would personally prefer if we used `member` instead of `key` throughout this article, since that's how we refer to this in other articles we've written. Consistency should help readers understand what these things are. Additionally, this is also the spec vocabulary.

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+
+### Example solution
+
+As the example application is a single page, we can use `"/"`, or omit the key altogether. For that same reason, we can display cycleTracker without browser tools, setting the `display` to `standalone`.
```
```suggestion
As the example application is a single page, we can use `"/"` as the `start_url`, or omit the member altogether. For that same reason, we can display the app without the browser UI by setting the `display` to `standalone`.
```

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+As the example application is a single page, we can use `"/"`, or omit the key altogether. For that same reason, we can display cycleTracker without browser tools, setting the `display` to `standalone`.
+
+In [our CSS](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/HTML_and_CSS#css_file), the `background-color: #efe;` is set on the `body` element selector. We use `#eeffee` to ensure a smooth transition from placeholder appearance to app load.
+}
```
```suggestion
```

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+
+## App iconography
+
+PWA icons help users identify your app, make it more visually appealing, and improve discoverability. The PWA app icon appears on home screens, app launchers, browser address bars, in app store search results: When users search for apps in the app store, the PWA icon will be displayed in the search results. The size of the rendered icon and the file requirements varies depending on where it is di ...
```
```suggestion
PWA icons help users identify your app, make it more visually appealing, and improve discoverability. The PWA app icon appears on home screens, app launchers, or app store search results. The size of the rendered icon and the file requirements varies depending on where it is displayed and by whom. The manifest is where you define your images.
```

Just a suggestion to remove the address bar because the favicon is used there, not the manifest icon (afaik).

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
````diff
@@ -0,0 +1,229 @@
+```js
+{
+  "name": "MyApp",
+  "icons: [
````
```suggestion
  "icons": [
```

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
````diff
@@ -0,0 +1,229 @@
+}
+```
+
+The `purpose` keys should be set to `maskable` defining [icona as adaptive](https://web.dev/maskable-icon/).
````
Hm, I'm having trouble parsing this sentence. 
Also, not all icons need to be maskable. And we might want to link to https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/How_to/Define_app_icons#support_masking for more information on this topic.

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+
+The `purpose` keys should be set to `maskable` defining [icona as adaptive](https://web.dev/maskable-icon/).
+
+All icons should have the same look and feel to ensure users recognize your PWA, but the larger the icon, the greater the detail it can contain. All icon files are squares. Include safe zones so that when the image is masked by the operating system, it renders okay as a circle.
```
It might make sense to expand a little bit on what the safe zone is. We can borrow some of the text from that "support masking" section I mentioned before if needed.

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+
+The manifest file extension can be the specification suggestion `.webappmanifest`. However, being a JSON file, it is most commonly saved with the browser-supported `.json` extension.
+
+PWAs require a manifest file. We have a fully functional app, but it's not yet a PWA. We have to add the manifest the app. The web manifest is an external JSON file. To include the external JSON resource, the `rel="manifest"` is used. The `href` attribute of the `<link>` points to the location of the resource.
```
```suggestion
PWAs require a manifest file to be linked from the app's HTML document. We have a fully functional app, but it's not yet a PWA because it doesn't link to our external manifest JSON file yet. To include the external JSON resource, we use the `<link>` element, with the `rel="manifest"` attribute, and setting the `href` attribute to the location of the resource.
```

Just a suggestion to improve the flow of this paragraph, which seemed a bit hard to follow at first read.

captainbrosset commented on files/en-us/web/progressive_web_apps/tutorials/intro/manifest_file/index.md:
```diff
@@ -0,0 +1,229 @@
+
+## Up next
+
+Now that we have a PWA manifest file, let's [create a secure connection](/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro/Secure). While no frameworks are needed to create a PWA, we will use X to create a secure connection; a PWA requirement.
```
Curious what you have in mind with framework X in this sentence?
Note that when you're working on localhost (as most devs will do), there is no need for a secure connection. The browser will allow installing the PWA on http if the domain is localhost.

captainbrosset reviewed (commented):
Some more comments. I hope they're useful. Thanks for the work on this!

estelle committed:
app identity

estelle committed:
Apply suggestions from code review

Co-authored-by: Patrick Brosset <patrickbrosset@gmail.com>

estelle committed:
new image

estelle commented on files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/html.jpg:
i added a new image: screen shot from MS Edge ;)

estelle committed:
keys to members

estelle committed:
Update files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/index.md

Co-authored-by: Patrick Brosset <patrickbrosset@gmail.com>

estelle commented on files/en-us/web/progressive_web_apps/tutorials/intro/html_and_css/index.md:
```diff
@@ -0,0 +1,293 @@
+
+## HTML content
+
+Even if the HTML in `index.html` is familiar to you, we recommend reading thru this section before adding the [placeholder data](en-US/Docs/Web/Progressive_web_apps/Tutorials/Intro/HTML_and_CSS#placeholder_text) and [`styles.css`](en-US/Docs/Web/Progressive_web_apps/Tutorials/Intro/HTML_and_CSS#css-file), and creating `app.js`, the [application's JavaScript](/en-US/docs/Web/Progressive_web_apps/T ...
```
this is kind of the table of contents for this page, but reworded it to "adding some temporary hard-coded data"

estelle committed:
more newby lingo

estelle committed:
Update files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md

Co-authored-by: Patrick Brosset <patrickbrosset@gmail.com>

estelle commented on files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md:
```diff
@@ -0,0 +1,351 @@
+
+When a user visits the page, we check if they have existing data stored in local storage. The first time a user visits the page, there won't be any data. When a new user selects two dates and submits the form, we need to:
+
+1. Create a "`<h2>Past periods</h2>`" header
+2. Create an {{HTMLelement("ul")}}
```
i debated doing this. It was easier to add a line of JS then to explain the UX design requirements of empty data along with the CSS for the two different experiences. (the "empty" message would not be a list item semantically)

estelle committed:
checkDatesInvalid

estelle commented on files/en-us/web/progressive_web_apps/tutorials/intro/javascript_functionality/index.md:
```diff
@@ -0,0 +1,351 @@
+  if (!startDate || !endDate || startDate > endDate) {
+    // For now, we clear the dates if either
+    // or both are invalid
+    newPeriodFormEl.reset();
```
I understand and agree with what you are saying. If we were creating a robust app, this would not reset the form, but rather do the steps explained in the comments. That is why this was created as a separate function -- for the comments, not the current code. I reordered the comments in the function to make this clearer.

estelle committed:
icon masking

estelle commented:
> Some more comments. I hope they're useful. Thanks for the work on this!

Thanks for the excellent feedback. I addressed all the issues (I think), resolving those where I followed your suggestions, which were all but 2: 1) the checkDateInvalid is indeed not a good function. I kept it as a place to hold comments to explain what would be required for a better function with suggestions for better validation, and 2) i kept the h2 in the JS, keeping the original div empty when there is no content, to avoid having to use an `li` when a `p` is appropriate, to not to have to explain UI of empty data, and to avoid have to write and explain CSS associated with that UI.

captainbrosset commented:
@estelle the content looks great, but I'm confused about one thing that I'd like to understand before approving this: when I go to the preview site for this PR, I can see a bunch of content that's not in the PR, like this intro page: https://pr26891.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro
Or the changes you made to the tutorial main page: https://pr26891.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Tutorials

Do you know where this content comes from?

I also see a lot of red links in the preview and I don't know if you intend to add the remaining pages in this PR or later.
Personally, I'm happy for this to be merge, knowing that it's already been a lot of work, and then merging follow-up PRs to finish everything up.

Let me know.

estelle commented:
There are two PRs that are not ready for merge: 

the [/en-US/docs/Web/Progressive_web_apps/Tutorials](https://pr26891.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Tutorials) PR can be reviewed and approved, but should not be published until the tutorial is complete. That's the landing page for all the PWA tutorials. 

The [/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro](https://pr26891.content.dev.mdn.mozit.cloud/en-US/docs/Web/Progressive_web_apps/Tutorials/Intro) is the landing page for this tutorial. It won't be finished until the tutorial is finished as the overview of something keeps getting edited until that something is complete. That is in draft mode and is there for reviewers to understand the scope of the project and what missing sections are actually TBW.

I broke the project into a few PRs because, as you'll note, this PR already has 100 or so commits and over 50 comments. A whole project PR would be overwhelming.


> 
> Do you know where this content comes from?
> 
> I also see a lot of red links in the preview and I don't know if you intend to add the remaining pages in this PR or later. Personally, I'm happy for this to be merge, knowing that it's already been a lot of work, and then merging follow-up PRs to finish everything up.
> 
> Let me know.

estelle committed:
typo

estelle committed:
secure and manifest file names. Manifest example solution

estelle committed:
icon directory and favicon info

estelle committed:
typo
